Mathlib/CategoryTheory/Functor/TwoSquare.lean

```diff
@@ -0,0 +1,156 @@
+@[simps!]
+def vId (T : C₁ ⥤ C₂) : TwoSquare T (𝟭 _) (𝟭 _) T :=
+  𝟙 _
+
+/-- Notation for the vertical identity 2-square. -/
+scoped notation "𝟙ᵥ" => vId  -- type as \b1\_v
+
+/-- Whiskering a 2-square with a natural transformation at the top. -/
+@[simps!]
+protected def whiskerTop {T' : C₁ ⥤ C₂} (α : T ⟶ T') (w : TwoSquare T' L R B) : TwoSquare T L R B :=
+  whiskerRight α R ≫ w
+
+/-- Whiskering a 2-square with a natural transformation at the left side. -/
+@[simps!]
+protected def whiskerLeft {L' : C₁ ⥤ C₃} (α : L ⟶ L') (w : TwoSquare T L R B) :
+    TwoSquare T L' R B :=
+  w ≫ whiskerRight α B
+
+/-- Whiskering a 2-square with a natural transformation at the right side. -/
+@[simps!]
+protected def whiskerRight {R' : C₂ ⥤ C₄} (α : R ⟶ R') (w : TwoSquare T L R' B) :
+    TwoSquare T L R B :=
+  whiskerLeft T α ≫ w
+
+/-- Whiskering a 2-square with a natural transformation at the bottom. -/
+@[simps!]
+protected def whiskerBottom {B' : C₃ ⥤ C₄} (α : B ⟶ B') (w : TwoSquare T L R B) :
+    TwoSquare T L R B' :=
+  w ≫ whiskerLeft L α
+
+variable {C₅ : Type u₅} {C₆ : Type u₆} {C₇ : Type u₇} {C₈ : Type u₈}
+  [Category.{v₅} C₅] [Category.{v₆} C₆] [Category.{v₇} C₇] [Category.{v₈} C₈]
+  {T' : C₂ ⥤ C₅} {R' : C₅ ⥤ C₆} {B' : C₄ ⥤ C₆} {L' : C₃ ⥤ C₇} {R'' : C₄ ⥤ C₈} {B'' : C₇ ⥤ C₈}
+
+/-- The horizontal composition of 2-squares. -/
+@[simps!]
+def hComp (w : TwoSquare T L R B) (w' : TwoSquare T' R R' B') :
+    TwoSquare (T ⋙ T') L R' (B ⋙ B') :=
+  (Functor.associator _ _ _).hom ≫ (whiskerLeft T w') ≫
+  (Functor.associator _ _ _).inv ≫ (whiskerRight w B') ≫ (Functor.associator _ _ _).hom
+
+/-- Notation for the horizontal composition of 2-squares. -/
+scoped infixr:80 " ≫ₕ " => hComp -- type as \gg\_h
+
+/-- The vertical composition of 2-squares. -/
+@[simps!]
+def vComp (w : TwoSquare T L R B) (w' : TwoSquare B L' R'' B'') :
+    TwoSquare T (L ⋙ L') (R ⋙ R'') B'' :=
+  (Functor.associator _ _ _).hom ≫ (whiskerRight w R'') ≫
+  (Functor.associator _ _ _).inv ≫ (whiskerLeft L w') ≫ (Functor.associator _ _ _).hom
+
+/-- Notation for the vertical composition of 2-squares. -/
+scoped infixr:80 " ≫ᵥ " => vComp -- type as \gg\_v
+
+section Interchange
+
+variable {C₉ : Type u₉} [Category.{v₉} C₉] {R₃ : C₆ ⥤ C₉} {B₃ : C₈ ⥤ C₉}
+
+/-- When composing 2-squares which form a diagram of grid, compositing horionzall first yields the
+same result as composing vertically first. -/
+lemma hCompVCompHComp (w₁ : TwoSquare T L R B) (w₂ : TwoSquare T' R R' B')
+    (w₃ : TwoSquare B L' R'' B'') (w₄ : TwoSquare B' R'' R₃ B₃) :
+    (w₁ ≫ₕ w₂) ≫ᵥ (w₃ ≫ₕ w₄) = (w₁ ≫ᵥ w₃) ≫ₕ (w₂ ≫ᵥ w₄) := by
+  unfold hComp vComp
+  unfold whiskerLeft whiskerRight
+  ext c
+  simp only [Functor.comp_obj, NatTrans.comp_app, Functor.associator_hom_app,
+    Functor.associator_inv_app, comp_id, id_comp, Functor.map_comp, assoc]
+  slice_rhs 2 3 =>
+    rw [← Functor.comp_map _ B₃, ← w₄.naturality]
+  simp only [Functor.comp_obj, Functor.comp_map, assoc]
```

Comment:
It is unlikely that this would break in the future, but it is recommended not to squeeze fast terminal simps:
```suggestion
  simp
```